File: src/module/classify/13/513/2641.py
class Solution:
    def replaceValueInTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        root.val = 0
        q = [root]
        while q:
            tmp = q
            q = []

            # 计算下一层的节点值之和
            next_level_sum = 0
            for node in tmp:
                if node.left:
                    q.append(node.left)
                    next_level_sum += node.left.val
                if node.right:
                    q.append(node.right)
                    next_level_sum += node.right.val

            # 再次遍历，更新下一层的节点值
            for node in tmp:
                children_sum = (node.left.val if node.left else 0) + \
                               (node.right.val if node.right else 0)
                if node.left: node.left.val = next_level_sum - children_sum
                if node.right: node.right.val = next_level_sum - children_sum
        return root

File: src/module/tree/2_13/2641.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var replaceValueInTree = function (root) {
  let queue = [];
  if (root != null) {
    queue.push(root);
    root.val = 0;
  }
  while (queue.length) {
    let nextLevelSum = 0;
    let size = queue.length;
    let nextQueue = [];
    for (let i = 0; i < size; i++) {
      let node = queue[i];
      if (node.left) {
        nextQueue.push(node.left);
        nextLevelSum += node.left.val;
      }
      if (node.right) {
        nextQueue.push(node.right);
        nextLevelSum += node.right.val;
      }
    }
    for (let i = 0; i < size; i++) {
      let node = queue[i];
      let childrenSum =
        (node.left ? node.left.val : 0) + (node.right ? node.right.val : 0);
      if (node.left) {
        node.left.val = nextLevelSum - childrenSum;
      }
      if (node.right) {
        node.right.val = nextLevelSum - childrenSum;
      }
    }
    queue = nextQueue;
  }
  return root;
};
